From a62ccc28271072ffbc19530bce2356d7e4bca033 Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Mon, 5 Apr 2021 13:44:33 +0300
Subject: [PATCH backport 4.19 2/2] mlxsw: minimal: Add support for setting
 line card device address

Extend structure 'mlxsw_m' with 'max_modules_per_slot' field.
Use this field for setting address for device found on line card at
slot #n. The device address will be set with slot #n related offset.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/minimal.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/minimal.c b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
index 093989625cd0..604daaa1847c 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/minimal.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
@@ -30,6 +30,7 @@ struct mlxsw_m {
 	struct mlxsw_m_area *main;
 	struct mlxsw_m_area **linecards;
 	u8 max_ports;
+	u8 max_modules_per_slot;
 };
 
 struct mlxsw_m_area {
@@ -123,6 +124,10 @@ mlxsw_m_port_dev_addr_get(struct mlxsw_m_port *mlxsw_m_port)
 	 * value.
 	 */
 	dev->dev_addr[ETH_ALEN - 1] = mlxsw_m_port->module + 1;
+	/* Add MAC address offset for slot. */
+	if (mlxsw_m_port->slot_index)
+		dev->dev_addr[ETH_ALEN - 1] += (mlxsw_m_port->slot_index - 1) *
+					       mlxsw_m->max_modules_per_slot;
 	return 0;
 }
 
@@ -239,8 +244,13 @@ static int mlxsw_m_ports_create(struct mlxsw_m_area *mlxsw_m_area, u8 slot_index
 	if (err)
 		return err;
 
-	mlxsw_reg_mgpir_unpack(mgpir_pl, NULL, NULL, NULL,
-			       &mlxsw_m_area->max_ports, NULL, NULL);
+	if (slot_index)
+		mlxsw_reg_mgpir_unpack(mgpir_pl, NULL, NULL, NULL,
+				       &mlxsw_m_area->max_ports, NULL, NULL);
+	else
+		mlxsw_reg_mgpir_unpack(mgpir_pl, NULL, NULL, NULL,
+				       &mlxsw_m_area->max_ports, NULL,
+				       &mlxsw_m_area->mlxsw_m->max_modules_per_slot);
 
 	if (!mlxsw_m_area->max_ports)
 		return 0;
-- 
2.11.0

